### ISSUE TITLE
[Feature Request] - RTE Session - Dragging .FBX to scene results in Entity populated with multiple mesh components.

### ISSUE THREAD

**Issue by shawstar** [o3de/o3de#9885]
**Describe the bug**

"After user adding the RV mesh into the O3DE scene, the RV entity instantiated with two Mesh components. This isn't allowed, so the second mesh component was disabled due to conflict with first mesh. Possible issue - why were there two mesh components in the first place? This happened again when loading another model that has multiple models within it. This was a more complex model, so when loading it to the Editor, the entity contains 4+ more Mesh components than needed."

I did a repo of this issue and duplicated the results. 

**Steps to reproduce**
1. Drag in an .FBX with 1 mesh node.
2. This will add more than 1 Mesh Component to the Entity.

![image](https://user-images.githubusercontent.com/87207603/171674006-34c47a92-6c23-40b5-8509-49c4eec34d11.png)

**Expected behavior**

![image](https://user-images.githubusercontent.com/87207603/171674492-68def264-684b-4a42-af23-f602b10c7229.png)
Above is the source mesh structure, there is just 1 mesh.

**Actual behavior**
Adding static meshes, and even meshes with bones (Actors) will result in adding many mesh components even when only 1 mesh is in the .fbx file.

**Assets required**
Just dragging a simple box .fbx from Blender will produce this:
![image](https://user-images.githubusercontent.com/87207603/171675210-ffe86c42-9f44-4d9d-94fc-1c94a3b46088.png)
Here is the test file:
[SimpleBox.zip](https://github.com/o3de/o3de/files/8825671/SimpleBox.zip)

**Found in Branch**
Nightly 6-2-2022

**Desktop/Device (please complete the following information):**
 - PC
 - OS: Windows
 - Version 10
 - CPU Intel I9-9900k
 - GPU NVidia RTX 2080ti
 - Memory [e.g. 16GB]

**Additional context**

**Comment by lmbr-pip** [comment]
Relates to https://github.com/o3de/o3de/issues/9344 - needs UX changes to support this work.

**Comment by yuyihsu** [comment]
Convert this ticket to feature request, per conversation with @AMZN-stankowi

**Comment by yuyihsu** [comment]
This issue will show when a fbx file has multiple mesh groups in it.

Chatted with @AMZN-stankowi , here's our conclusion:
1. for the .FBX file with joints in it-
    - If user drag and drop the .fbx file to the scene, an actor component should be auto added.
    - If user drag and drop the .procprefab to the scene, the ".actor" file should be auto-added to the procprefab in the outliner **_(this is not currently supported today, need the prefab team to plan this in their roadmap)_**
2. for the regular .fbx (non-actor) file with multiple mesh groups
    - If user drag and drop the .fbx file to the scene, it should auto-generate the procprefab in the outliner. Within the procprefab, each mesh group will be shown as an individual entity, and each entity will have just one mesh component assigned to it. **_(this is not currently supported today, need the prefab team to plan this in their roadmap)_**
    - If user drag and drop the procprefab to the scene, it should show the same behavior as above. **_(this is not currently supported today, need the prefab team to plan this in their roadmap)_**

**Comment by rainbj** [comment]
Hi @shawstar, @yuyihsu, @monroegm, @lmbr-pip, @AMZN-stankowi, @srikappa-amzn, @AMZN-daimini 

After our meeting today, I'm going to summarize the approach that seems to be the most comparable to how a good phased approach fix could work here. 

- First, this fix should be going into the Sept release. While the long term fix will be going into the end of Q4 release (date yet to be specified). That long term fix we plan on using prefab overrides and all of its awesomeness. Until then we need an interment plan. 
- So initially we planned to fix this issue by; when a user drags in an FBX file, we would actually instantiate a procprefab file in the Outliner for them. The problem here is that a procprefabs main purpose it to maintain its inheritance from their DCC tool. Allowing you to update your FBX file and for O3DE to automatically pick up those changes.  This sounds good, EXCEPT this workflow prevents a user from being able to edit this file unless they implicitly know how to detach a procprefab AND they would have to do it every time. 
- As a group, we believe each of these use cases is SOMETIMES true. However, to maintain editability use case might not be the primary use case for most users. We believe that most user want the ability to drag in an FBX object and maintain the ability to perform edits on their FBX without having to learn to convert their objects back to an entity. 
- **Conclusions** is a compromise between the two workflows. We believe giving a user a modal dialog that specifies these two options is the right choice. First option, would say (and I'm summarizing, so please know the wording might need a few passes to get right) "Allow editing on my FBX file but loose preserving file syncing"  OR  "Preserve file Sync with FBX but file editing not allowed." 
- We should have a checkbox allowing users to save this choice, but it should also be modifiable in the setting.
- If the user chooses to import with editability, we take the procprefab and then automatically do a detach operation without the customer having to understand the underlying process.
- If the user choose to import with preserving syncing we just bring in a procprefab. User can do a detach operation later if they choose to do so. Learning of this operation may take new users time to understand.
-
![Dependent Assets](https://user-images.githubusercontent.com/82551958/177877071-efb9ee0b-d241-41d0-b5a0-a1885048963e.png)

**Comment by AMZN-daimini** [comment]
> This sounds good, EXCEPT this workflow prevents a user from being able to edit this file unless they implicitly know how to detach a procprefab AND they would have to do it every time.

Small correction, detaching removes the tie between the instances in the editor and the file, so either way you're not editing the file.

>  First option, would say (and I'm summarizing, so please know the wording might need a few passes to get right) "Allow editing on my FBX file but loose preserving file syncing" OR "Preserve file Sync with FBX but file editing not allowed."

The first and default option should be preserving the sync in my opinion, since it's the preferred workflow in the long term (when overrides are introduced)

**Comment by lawsonamzn** [comment]
Yeah, once overrides are supported the obvious choice is always to just instantiate a new user-side prefab that has the procprefab as an instance in it.  Then you get to edit things AND you still keep sync.

However, I'm not going to be able to implement that immediately (I'm working on some basic drag and drop APIs).  I'll keep those workflows in mind while working on the APIs but the initial one is likely to be very simple 
EG:

 you drag any asset (not just fbx) that produces a procprefab -> it instantiates the procprefab.  More than 1 procprefab?  It instantiates them all.   (You can always delete them or pick individual ones to drag).
 You drag a prefab -> instantiates prefab
 You drag a fbx which has no prefabs or procprefabs or actors --> unclear what to do?

My current plan is to make it so that gems that provide assets (actors, meshes, procprefabs...) can listen to the drag and drop events and choose to respond / consume the message and handle the drop if its a type they care about.  So the actor drop handler will handle the case where a fbx that produces an actor is present.  A procprefab handler will handle the case where the FBX produces a procprefab.  
Each handler has a priority so UX can choose what order to put them such that one or the other 'wins' when theres actors, meshes, procprefabs, materials, and other stuff in a fbx...

**Comment by lawsonamzn** [comment]
To be more specific here, currently there's just one big handler for all drag and drop, the AzAssetBrowserEventHandler class, and that one class tries to do its best based on querying a whole range of other busses to figure out what to do.

But that pattern falls apart here, because it turns out each different type of asset needs special handling that is generally specific to whatever component understands that asset, to make it work in the scene.  The current default behavior of finding whatever component accepts that kind of assets and just slapping one of those on an entity for every such selected asset clearly doesn't work how people expect.

**Comment by lawsonamzn** [comment]
Hmm, just found out that QT actually adds 2 of the same element to the mimedata during drag.  This might be because the asset browser has 2 "columns" in its view, and the qt element puts both columns in the drag operation

**Comment by rainbj** [comment]
Hi Nicholas, I just wanted to let you know @yuyihsu and I are setting up a meeting on Monday with you to talk through this issue. We had a high level conversation with @lmbr-pip and @AMZN-stankowi today, and UX has some concerns about the use of procprefab in all workflows for FBX drag and drop. Mostly around instantiating a procprefab on the most basic FBX when the user intentions are to add components to those basic entities to perform actions. Example, I want to add a fence post FBX and then add an entity with physX collision component. If  we go down the procprefab road, we would have to force users to learn about detaching a procprefab because procprefab do not allow any modification to itself. This means to add any components, a user would first have to do detach. The learning curve for the user to overcome is too much for a new engine and user, and would highly suggest we consider carefully the implications of this mental load on users.  Workflows for doing the most basic editing, and in those instances I am against using Procprefab as the state the user sees. I've added a screenshot above that show us asking the user what their intentions are at this moment with this asset. It's unfortunate that we can't predict this behavior. However, assuming that all people have a complex asset is also too far in the other direction. When initially coming up with procprefab its specific use case for creation was. As a user, I want this specific asset to maintain the assets properties coming from the DCC tool. However, this was with the knowledge of overrides working with prefabs. Adding a new prefab override component is still a ways off from being developed. However, this bug is visible now and should be addressed in a way that address the issue for our engine as it currently stands and not something that will be in the future. Once overrides comes in, and we can test how it works with Procprefabs then we adjust the workflow to match the expectation

https://user-images.githubusercontent.com/82551958/180519799-dd81fb44-53ee-4f88-8f96-5fdc32510103.mp4

.

**Comment by lawsonamzn** [comment]
Yes, the way I see it there's some short term fixes we need to apply that we can do immediately and quickly (low cost) to fix the workflow bugs like this issue. 

Quickly fix the problems it causes right now, and then plan some some mid-to-longer term approaches we can take.   

For example, adding multiple mesh components to the same entity is not OK, its definitely a bug, since those components don't work on the same entity.  So a quick fix I'm applying is to just create one such mesh component, the one that most likely represents the fbx being dragged.  

Note that even if we make plans that involve proc prefabs, we still have to have a heuristic that works when there isn't a procprefab.  For example, we have to handle all of these situations

* FBX with procprefab, actor, 8 meshes, 4 materials, is dragged to the viewport
* FBX with no proc prefab, but actor, 8 meshes, 4 materials, is dragged to the viewport
* FBX with 8 meshes, 4 materials, is dragged to the viewport
* FBX with 1 meshes, 0 materials is dragged to the viewport
* FBX with 0 meshes, 1 materials is dragged to the viewport
* FBX with 0 meshes, 4 materials is dragged to the viewport
* FBX with 4 actors is dragged to the viewport
* User expands the FBX and drags individual components (procprefab, material, actor) to the viewport
* User does all of the above, but instead of on the viewport, drops it on the outliner ontop of another entity.
* User does all of the above but instead of the viewport, drops it on the outliner into a prefab thats got focus or does not have focus

in addition, there's also an important reminder that o3de asset types themselves are in gems (plugin addons) - for example the Atom gem is what creates the existence of the Model or Material assets.  Its possible for a user developing a game to create a new type of asset, for example "foo asset" or "bar asset", and have the FBX export those new types of assets too . The system needs to be able to have them put code in their gem to handle what happens when an FBX contains a Foo asset and drops it into the viewport, which infers that the approach has to also be modular, we can't make one big decision tree for all the types that ever will exist and put it in a central code location.

**Comment by rainbj** [comment]
Joseph brought up a good point about we have one missing use case for the error case, and we present a modal pop up. If you could present a use case, I can give a design for this use case.

**Comment by lawsonamzn** [comment]
I gave some of the scenarios above.  All of those scenarios are currently possible to happen.

**Comment by lawsonamzn** [comment]
Also bear in mind that there are ways to achieve all user intents:
* User can drag the fbx itself to the viewport
* User can **expand** the fbx, then drag individual pieces to the viewport -> spawn new entity with that component.
* User can **expand** the fbx, then drag individual pieces to the inspector with an entity already being inspected (add component to existing entity)

In the first case, I read the user's intent as "just make me the thing in the viewport that corresponds to this logical fbx"
In the second case, I read the user's intent as "I want this specific aspect of it to be created in the viewport"
In the third case, I read the users intent as "I want to add this specific part to this existing entity."

As such, do we really need additional, potential extra-click or complexity-adding modal popups to clarify intent?

Expanding the thing and dragging a piece is clear and unambiguous.  The only remaining ambiguity is if the user drags the entire container asset (fbx or other) into the viewport.  But even then, the user's _intention_ is clear.  They want whatever that file logically represents (the actor, the 'thing', etc) to become real in the viewport.  We probably don't need confirmation of their intention there either, but we need to figure out how to achieve that intention...
* when there is a proc prefab (probably spawn the proc prefab since that best represents the thing)
* when there is no proc prefab (now its difficult to answer since unless there's just 1 mesh, its hard to guess what 'thing' inside the fbx represents it best).

**Comment by AMZN-stankowi** [comment]
the workflow I was trying to cover was: You drag and drop the source FBX asset, and the FBX's only product assets do not have a logical instantiation target. The FBX either has no product assets, or it only has product assets that we can't/don't spawn (just animation output, just buffers, or just physics meshes). I was suggesting in that case, it would feel bad to drag and FBX and have nothing happen, so instead a popup might be more desirable "You tried to drag empty.fbx into a scene and it has no product assets, so we can't do anything here, sorry."

**Comment by lawsonamzn** [comment]
Well the current behavior shows the "No Entry" sign by default by telling the operating system that there's nothing in the data that can be dropped there.  (The operating system changes the mouse cursor to the no-drop cursor).  

It is important to tell the OS this and do it this way, becuase other technologies that are part of the operating system, such as clipboard or mimedata managers, but also assistive technologies such as screen readers or other alternate input devices for those with a disability need to know about available drop targets and so on.

While it would be possible to do as you say, it would involve telling the operating system that you can in fact handle the data being dragged even when you can't actually, and then respond to the operating system's request to drop the data by showing such a message.  This would affect how the OS "sees" the various widgets and windows in terms of drag an drop and assistive technologies that can pick up a thing and tab to the next droppable widget for example would consider your drop target a "stop".

It feels like a better option is to just return the "I can't handle this kind of data" return option, which will cause the OS to show the "you can't drop that here" icon in places you can't drop it.

**Comment by yuyihsu** [comment]
@lmbr-pip , can you help split the work in this ticket?

**Comment by lawsonamzn** [comment]
Note that the above PR isn't going to 100% solve all problems with dragging stuff around.  But at the very least it should stop dragging stuff causing duplicate mesh components unless you intentionally drag from the asset browser into the inspector area of the entity.   (In which case, you chose to add all those components to the selected entity, intentionally)

**Comment by lawsonamzn** [comment]
I'd recommend adding additional tickets for e.g. improving the way we generate procprefabs (ie, making it possible and easy to generate default procprefabs even with scene settings present), as well as the quality of the generated procprefab (they currently think each lod is a different mesh)

**Comment by lawsonamzn** [comment]
I committed the first wave of drag and drop improvements.  I'm not sure it address _everything_ in this Issue, but perhaps this should be split into pieces of achievable work and then closed?

**Comment by amzn-leenguy** [comment]
UX to confirm to resolve this ticket. Please feel free to create new tickets for any other requested changes mentioned in this thread. (If you want you can add comments here to relate the tickets)

**Comment by AMZN-stankowi** [comment]
@amzn-leenguy why did you re-open this ticket? The last few messages on this thread mention this ticket is completed.

**Comment by lawsonamzn** [comment]
This kind of thing can always use improvement but perhaps create a new ticket for the specific improvements.
